File: apps/web/src/lib/comment-threads.ts
import type { Comment } from "@/db/schema"

/**
 * EXP-741: one issue's comments folded into threads — the top-level cards in
 * their input order, and each card's replies (also in input order) keyed by
 * the parent id. Threads are ONE level deep by construction (`comments.create`
 * re-parents a reply-to-a-reply onto the root), so a row is a reply exactly
 * when `parentId` is set.
 *
 * A reply whose parent is NOT in the list (still syncing, or the parent row
 * vanished from a partial snapshot) surfaces as a top-level card rather than
 * disappearing — the row is still real activity. Mirrored by the desktop
 * `thread_comments`, iOS `threadComments` and Android `threadComments`.
 */
export interface CommentThreads<C> {
  topLevel: C[]
  repliesByParent: Map<string, C[]>
}

export function threadComments<C extends Pick<Comment, `id` | `parentId`>>(
  comments: readonly C[]
): CommentThreads<C> {
  const ids = new Set(comments.map((comment) => comment.id))
  const topLevel: C[] = []
  const repliesByParent = new Map<string, C[]>()
  for (const comment of comments) {
    const parentId = comment.parentId
    if (parentId && ids.has(parentId) && parentId !== comment.id) {
      const list = repliesByParent.get(parentId) ?? []
      list.push(comment)
      repliesByParent.set(parentId, list)
    } else {
      topLevel.push(comment)
    }
  }
  return { topLevel, repliesByParent }
}

/** How many rows a thread list holds in total — the "Activity (N)" count. */
export function countThreadedComments<C>(threads: CommentThreads<C>): number {
  let replies = 0
  for (const list of threads.repliesByParent.values()) replies += list.length
  return threads.topLevel.length + replies
}

File: apps/ios/ExpCore/Sources/Domain/CommentThreads.swift
import Foundation

/// EXP-741: one issue's comments folded into threads — the top-level cards in
/// their input order, and each card's replies (also in input order) keyed by
/// the parent id. Mirrors web `lib/comment-threads.ts`, the desktop
/// `thread_comments` and Android `threadComments`.
public struct CommentThreads: Sendable {
    public var topLevel: [CommentEntity]
    public var repliesByParent: [String: [CommentEntity]]

    public init(topLevel: [CommentEntity] = [], repliesByParent: [String: [CommentEntity]] = [:]) {
        self.topLevel = topLevel
        self.repliesByParent = repliesByParent
    }

    /// Every row — the "Activity (N)" count.
    public var count: Int {
        topLevel.count + repliesByParent.values.reduce(0) { $0 + $1.count }
    }
}

/// Threads are ONE level deep by construction (`comments.create` re-parents a
/// reply-to-a-reply onto the root), so a row is a reply exactly when
/// `parentId` is set. A reply whose parent is NOT in the list (still syncing,
/// or gone from a partial snapshot) surfaces as a top-level card rather than
/// disappearing — the row is still real activity.
public func threadComments(_ comments: [CommentEntity]) -> CommentThreads {
    let ids = Set(comments.map(\.id))
    var threads = CommentThreads()
    for comment in comments {
        if let parentId = comment.parentId, parentId != comment.id, ids.contains(parentId) {
            threads.repliesByParent[parentId, default: []].append(comment)
        } else {
            threads.topLevel.append(comment)
        }
    }
    return threads
}
